### PR TITLE
tools: Add locking to pyboard.py serial port for linux and OSX

### PR DESCRIPTION
**Pull request by dhylands**
This adds proper locking for the serial port on linux and OSX.

If a program like picocom or minicom has the serial port open and you try to run pyboard, you'll now get an error message.

### CODE REVIEW

**Comment by pfalcon** [comment]
This adds whole bunch of code. Is it really *that* needed?

**Comment by dhylands** [comment]
Well, it adds a bunch of code on the PC side of things (which personally I'm perfectly happy with). I know I've run into the problem myself numerous times having rshell (which derives from pyboard.py) fail to connect to the pyboard becuase I had some other terminal program running in the background.

If this isn't accepted here, I'll just add it to rshell, but logically the functionality belongs here. Logically, it probably belongs in a separate file, but I know in the past @dpgeorge has preferred to 
keep pyboard.py as a single file and not add additional dependencies.

Every terminal program (piccom, minicom, screen, etc) has similar code.

So to summarize, it's not absolutely needed, but without it you get failures trying to run tests if you happen to have the repl open someplace else.

**Comment by pfalcon** [comment]
> a bunch of code on the PC side of things

Yes, which may have bugs, or not work as expected for someone, etc.

I'm +0.5 on this, but let @dpgeorge decide on the merge.

**Comment by dhylands** [comment]
I fixed a couple minor issues and tested under Python2 & 3 on Windows, Linux, and OSX.

**Comment by dpgeorge** [comment]
Running on Arch Linux I get the following error:
```
Traceback (most recent call last):
  File "./tools/pyboard.py", line 442, in <module>
    main()
  File "./tools/pyboard.py", line 382, in main
    pyb = Pyboard(args.device, args.baudrate, args.user, args.password, args.wait)
  File "./tools/pyboard.py", line 213, in __init__
    self.file_lock.lock()
  File "./tools/pyboard.py", line 162, in lock
    with open(self._lockname, 'w') as file:
PermissionError: [Errno 13] Permission denied: '/var/lock/LCK..ttyACM0'
```
Also, running picocom on the same device I don't see the LCK file... I assume that it's supposed to be a standard name that all programs create if they lock the device.

**Comment by dhylands** [comment]
Hmm. I'll have to install Arch in a VM and try it out. I actually got the filename from the picocom source files, but picocom supports several different methods of locking. It's possible that Arch decided to implement something different.

**Comment by dpgeorge** [comment]
This is a useful feature but I think it's better suited to integrate as part of #6375, and to only cover the case of multiple pyboard.py's trying to open the same port (ie don't also try to protect against minicom/picocom/etc).

**Comment by davehylands** [comment]
It looks like pyserial 3.3 added an exclusive argument to the __init__ method.

**Comment by dpgeorge** [comment]
> It looks like pyserial 3.3 added an exclusive argument to the **init** method.

Very nice!  See #7141.
